driver-core/src/test/functional/com/mongodb/operation/MixedBulkWriteOperationSpecification.groovy

```diff
@@ -540,22 +540,28 @@ class MixedBulkWriteOperationSpecification extends OperationFunctionalSpecificat
         [async, ordered] << [[true, false], [true, false]].combinations()
     }
 
+    @Category(Slow)
     def 'should split the number of writes is larger than the match write batch size'() {
         given:
+        def binding = async ? getAsyncSingleConnectionBinding() : getSingleConnectionBinding()
+        def maxWriteBatchSize = getCollectionHelper().isMaster().getInteger("maxWriteBatchSize").intValue()
+        def numberOfWrites = maxWriteBatchSize + 100
         def writes = []
-        (0..2000).each {
+
+        (1..numberOfWrites).each {
             writes.add(new InsertRequest(new BsonDocument()))
         }
-        def operation = new MixedBulkWriteOperation(getNamespace(), writes, ordered, ACKNOWLEDGED)
+        def operation = new MixedBulkWriteOperation(getNamespace(), writes, ordered, writeConcern)
 
         when:
-        execute(operation, async)
+        execute(operation, binding)
+        acknowledgeWrite(binding)
 
         then:
-        getCollectionHelper().count() == 2001
+        getCollectionHelper().count() == numberOfWrites + 1
 
         where:
-        [async, ordered] << [[true, false], [true, false]].combinations()
+        [async, ordered, writeConcern] << [[true, false], [true, false], [ACKNOWLEDGED, UNACKNOWLEDGED]].combinations()
```

Comment:
I think there's a bug in BulkWriteBatch, which I noticed while seeing if this test is a sufficient one for splitting.  The BulkWriteBatch constructor loops through all the items, and creates the first batch with all the write requests that are the same type as the first one.  But this is only correct if the writes are unordered.  Otherwise, the driver should not attempt to execute the write requests out of order.

Reply:
Good catch @jyemin - not sure when I regressed on that but have added a unit test for BulkWriteBatch to ensure it doesn't regress again.